plugins/pixart-rf/fu-pxi-device.c

```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (C) 2020 Richard Hughes <richard@hughsie.com>
+ * Copyright (C) 2020 Jimmy Yu <Jimmy_yu@pixart.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+#include "fu-pxi-device.h"
+#include "fu-pxi-device-ota.h"
+#include "fu-common.h"
+#include "fu-pxi-device-common.h"
+#include "fu-udev-device.h"
+#include <linux/hidraw.h>
+#include <sys/ioctl.h>
+#include <linux/input.h>
+#include <linux/uhid.h>
+
+
+#define PXI_HID_DEV_OTA_INPUT_REPORT_ID 0x05
+#define PXI_HID_DEV_OTA_OUTPUT_REPORT_ID 0x06
+#define PXI_HID_DEV_OTA_FEATURE_REPORT_ID 0x07
+
+G_DEFINE_TYPE (FuPxiDevice, fu_pxi_device, FU_TYPE_UDEV_DEVICE)
+
+
+static FuFirmware *
+fu_pxi_device_prepare_firmware (FuDevice *device,
+                                GBytes *fw,
+                                FwupdInstallFlags flags,
+                                GError **error)
+{
+    g_autoptr(FuFirmware) firmware = fu_firmware_new ();
+
+
+    guint32 idx = 0;
+
+    gsize fw_sz = g_bytes_get_size(fw);
+    guint8 pos = 0;
+    gchar version[5] = {'.', '.', '.', '.', '.'};
+    const guint8 *fw_ptr = g_bytes_get_data(fw, &fw_sz);
+
+    if (!fu_firmware_parse (firmware, fw, flags, error))
+        return NULL;
+
+    fu_common_dump_raw (G_LOG_DOMAIN, "fw last 32 bytes", &fw_ptr[fw_sz - 32], 32);
+
+    /* find the version tag */
+    for(idx = 0; idx < 32; idx++)
+    {
+        if(fw_ptr[(fw_sz - 32) + idx ] == 'v')
+        {
+            pos = idx;
+            break;
+        }
+    }
```

Comment:
please do all this in a FuPxiFirmware object and then fu_pxi_firmware_new() in the header

Reply:
You'll also need to use fu_plugin_add_firmware_gtype() for this new type too.